Fix #12412 Use enum type for discriminator

bernie-schelberg-mywave opened the pull request:
As described in #12412, it's reasonably common to define the valid options for a discriminator property as an enum. However, the current versions of the generator assume the discriminator property is always `string`. This change uses the specified type when it is explicitly defined as a separate type in the schema. I attempted to also handle the scenario where the enum property was specified inline, but it was too complicated for me.

<!-- Please check the completed items below -->
### PR checklist
 
- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

bernie-schelberg-mywave commented:
@cachescrubber (2022/02) @welshm (2022/02) @MelleD (2022/02) @atextor (2022/02) @manedev79 (2022/02) @javisst (2022/02) @borsch (2022/02) @banlevente (2022/02) @Zomzog (2022/09)

ddoming73 commented:
Any chance of this PR being merged soon?

welshm commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:
```diff
@@ -3385,8 +3385,16 @@ protected CodegenDiscriminator createDiscriminator(String schemaName, Schema sch
 
-        // FIXME: for now, we assume that the discriminator property is String
-        discriminator.setPropertyType(typeMapping.get("string"));
+        // FIXME: there are other ways to define the type of the discriminator property (inline
+        //  for example). Handling those scenarios is too complicated for me, I'm leaving it for
+        //  the future..
+        String propertyType =
+            Optional.ofNullable(schema.getProperties())
+                .map(p -> (Schema) p.get(discPropName))
+                .map(Schema::get$ref)
+                .map(ModelUtils::getSimpleRef)
+                .orElseGet(() -> typeMapping.get("string"));
+        discriminator.setPropertyType(propertyType);
```
This will impact more than just the JavaSpring generation - can you regenerate the samples and tests to see what impact this has on other clients?

bernie-schelberg-mywave replied:
@welshm I ran `./bin/generate-samples.sh` and it ran for a while, but there were no changes to any of the samples in the git repo. Is that all I need to do?

bernie-schelberg-mywave replied:
@welshm I've run the commands in the PR instructions above. No files changed to commit to Git. I'm not certain that it worked properly, but the scripts seemed to run without errors.

welshm commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:
```diff
@@ -3385,8 +3385,16 @@ protected CodegenDiscriminator createDiscriminator(String schemaName, Schema sch
+        //  the future..
+        String propertyType =
+            Optional.ofNullable(schema.getProperties())
+                .map(p -> (Schema) p.get(discPropName))
```
nit: I know this isn't a variable you named, but I think full naming is better to disambiguate. `discriminatorPropertyName` might be more clear what the variable clearly represents

bernie-schelberg-mywave replied:
Variable has been renamed, please review again

welshm reviewed (commented):
Code look great! Can you run the commands in the PR template to regenerate the samples?

@wing328 will need to approve running the tests

cachescrubber commented on modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml:
```diff
@@ -170,6 +170,37 @@ components:
+      properties:
+        fruitType:
+          $ref: "#/components/schemas/FruitTypeEnum"
+      required: fruitType
```
The required property must be an array.

bernie-schelberg-mywave replied:
@cachescrubber why do you say that? In the documentation here: https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/ all the examples have a required discriminator field of type `string`. I actually thought that I'd read somewhere that it must be string, but I can't find that reference now.

bernie-schelberg-mywave replied:
Sorry, I see the problem now. I've pushed a fix.

cachescrubber commented on modules/openapi-generator/src/test/resources/3_0/oneOfDiscriminator.yaml:
```diff
@@ -170,6 +170,37 @@ components:
+        - $ref: '#/components/schemas/BananaOneOfEnumMappingDisc'
+      discriminator:
+        propertyName: fruitType
+        mapping:
```
The mapping to the enum value might cause problems with the auto generated JsonSubType.

cachescrubber replied:
I actually generated a sample Project:

```
@JsonIgnoreProperties(
  value = "fruitType", // ignore manually set fruitType, it will be automatically generated by Jackson during serialization
  allowSetters = true // allows the fruitType to be set during deserialization
)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "fruitType", visible = true)
@JsonSubTypes({
  @JsonSubTypes.Type(value = AppleOneOfEnumMappingDisc.class, name = "APPLE"),
  @JsonSubTypes.Type(value = AppleOneOfEnumMappingDisc.class, name = "AppleOneOfEnumMappingDisc"),
  @JsonSubTypes.Type(value = BananaOneOfEnumMappingDisc.class, name = "BANANA"),
  @JsonSubTypes.Type(value = BananaOneOfEnumMappingDisc.class, name = "BananaOneOfEnumMappingDisc")
})

@Generated(value = "org.openapitools.codegen.languages.SpringCodegen", date = "2022-11-03T09:30:55.032940+01:00[Europe/Berlin]")
public interface FruitOneOfEnumMappingDisc {
    public FruitTypeEnum getFruitType();
}

```

@welshm Isn't there another issue / pr dealing with duplicated JsonSubTypes?

cachescrubber replied:
@bernie-schelberg-mywave Could you please add a enum based discriminator to Ione of the generated sample projects (for example `bin/configs/spring-boot-oneof.yaml`)?

ddoming73 replied:
> I actually generated a sample Project:
> 
> ```
> @JsonIgnoreProperties(
>   value = "fruitType", // ignore manually set fruitType, it will be automatically generated by Jackson during serialization
>   allowSetters = true // allows the fruitType to be set during deserialization
> )
> @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "fruitType", visible = true)
> @JsonSubTypes({
>   @JsonSubTypes.Type(value = AppleOneOfEnumMappingDisc.class, name = "APPLE"),
>   @JsonSubTypes.Type(value = AppleOneOfEnumMappingDisc.class, name = "AppleOneOfEnumMappingDisc"),
>   @JsonSubTypes.Type(value = BananaOneOfEnumMappingDisc.class, name = "BANANA"),
>   @JsonSubTypes.Type(value = BananaOneOfEnumMappingDisc.class, name = "BananaOneOfEnumMappingDisc")
> })
> 
> @Generated(value = "org.openapitools.codegen.languages.SpringCodegen", date = "2022-11-03T09:30:55.032940+01:00[Europe/Berlin]")
> public interface FruitOneOfEnumMappingDisc {
>     public FruitTypeEnum getFruitType();
> }
> ```
> 
> @welshm Isn't there another issue / pr dealing with duplicated JsonSubTypes?

I'm running openapi-generator version 6.2.0 and the above is the current behaviour when you add a mapping to the discriminator. Both the derived classes and the mapping values are added as JsonSubTypes.  

The documentations seems to indicate that this is the expected behaviour. 

> The mapping in the discriminator includes any descendent schemas that allOf inherit from self, any oneOf schemas, any anyOf schemas, any x-discriminator-values, and the discriminator mapping schemas in the OAS document AND Codegen validates that oneOf and anyOf schemas contain the required discriminator and throws an error if the discriminator is missing.

bernie-schelberg-mywave replied:
@cachescrubber Yes, there's another project for duplicated JsonSubTypes: https://github.com/OpenAPITools/openapi-generator/pull/13815

bernie-schelberg-mywave commented:
@welshm @cachescrubber @wing328 Sorry for the delay, I've managed to generate samples (I added an example using an enum mapping), please review again.

GavinRay97 commented:
Thanks a ton for this @bernie-schelberg-mywave, this will help us out a ton at Hasura with our public SDK types, which currently fail to generate properly due to having a lot of polymorphic types using enum discriminants.

We have a query language IR where we distinguish between types of operations like `binary_op`, `unary_op`, `binary_array_op`, and operators like `eq`, `neq`, `lte` etc.
- https://github.com/hasura/graphql-engine-mono/blob/main/dc-agents/dc-api-types/src/agent.openapi.json#L1225-L1261

This fails pretty miserably in OpenAPI Generator currently:

```json
      "Expression": {
        "discriminator": {
          "mapping": {
            "and": "AndExpression",
            "binary_arr_op": "ApplyBinaryArrayComparisonOperator",
            "binary_op": "ApplyBinaryComparisonOperator",
            "exists": "ExistsExpression",
            "not": "NotExpression",
            "or": "OrExpression",
            "unary_op": "ApplyUnaryComparisonOperator"
          },
          "propertyName": "type"
        },
        "oneOf": [
          {
            "$ref": "#/components/schemas/ExistsExpression"
          },
          // <SNIPPED>
         ]
        },
      "BinaryComparisonOperator": {
        "additionalProperties": true,
        "anyOf": [
          {
            "enum": [
              "less_than",
              "less_than_or_equal",
              "greater_than",
              "greater_than_or_equal",
              "equal"
            ],
            "type": "string"
          },
          {
            "type": "string"
          }
        ]
      },
```

bernie-schelberg-mywave commented:
@welshm @cachescrubber @wing328 is there anything else I can do to help move this PR forward?

wing328 commented:
cc @OpenAPITools/generator-core-team as the change impacts default codegen.

bernie-schelberg-mywave commented:
@wing328 thanks for running the workflows, I saw there was a failure, so merged the `master` branch and re-generated the samples. Now I have a CircleCI failure, but I don't believe that it's because of the changes I've made. The error appears to be caused by a Java version incompatibility, although it builds using version 1.8 on my machine. I'm unable to run the `samples.circleci` profile on my local (`arm64e`-powered) machine. Can you provide any clue as to what I can do to resolve the failure?

wing328 commented:
Hi @bernie-schelberg-mywave thanks again for the PR.

Can you please PM me via Slack for some quick questions if you don't mind? 

ref: https://join.slack.com/t/openapi-generator/shared_invite/zt-12jxxd7p2-XUeQM~4pzsU9x~eGLQqX2g

nitish-atomic commented:
is there any timeline for having this in any release??

vojkozel96 commented:
Hi, do we have any progress here? We miss this functionality and it would help us a lot!

ludovicianul commented:
Is this planned to be merged? Seems like a key missing functionality? Is there a workaround for it that anyone is aware of (other than leaving the discriminator as String)?

maidi29 commented:
Any updates on this? 👀

ludovicianul commented:
As a workaround, I'm using a custom template for the `oneof_interface.mustache`.

bernie-schelberg-mywave commented:
As far as I know, the plan is to include this as part of the version 7.0 release.

bernie-schelberg-mywave commented on samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/APIs.swift:
```diff
@@ -5,14 +5,6 @@
 //
 
 import Foundation
-
-// We reverted the change of PetstoreClientAPI to PetstoreClient introduced in https://github.com/OpenAPITools/openapi-generator/pull/9624
-// Because it was causing the following issue https://github.com/OpenAPITools/openapi-generator/issues/9953
-// If you are affected by this issue, please consider removing the following two lines,
-// By setting the option removeMigrationProjectNameClass to true in the generator
-@available(*, deprecated, renamed: "PetstoreClientAPI")
-public typealias PetstoreClient = PetstoreClientAPI
-
```
I don't know why these changes are only in my branch, they appeared when I generated samples

bernie-schelberg-mywave commented on samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/APIs.swift:
```diff
@@ -31,8 +23,6 @@ open class RequestBuilder<T> {
     public let requiresAuthentication: Bool
 
     /// Optional block to obtain a reference to the request's progress instance when available.
-    /// With the URLSession http client the request's progress only works on iOS 11.0, macOS 10.13, macCatalyst 13.0, tvOS 11.0, watchOS 4.0.
-    /// If you need to get the request's progress in older OS versions, please use Alamofire http client.
```
I don't know why these changes are only in my branch, they appeared when I generated samples

bernie-schelberg-mywave commented on samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/Configuration.swift:
```diff
@@ -8,10 +8,6 @@ import Foundation
 
 open class Configuration {
     
-    // This value is used to configure the date formatter that is used to serialize dates into JSON format.
-    // You must set it prior to encoding any dates, and it will only be read once.
-    @available(*, unavailable, message: "To set a different date format, use CodableHelper.dateFormatter instead.")
-    public static var dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ"
```
I don't know why these changes are only in my branch, they appeared when I generated samples

bernie-schelberg-mywave commented on samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift:
```diff
@@ -52,7 +52,7 @@ open class URLSessionRequestBuilder<T>: RequestBuilder<T> {
      - retry the request.
      */
-    @available(*, deprecated, message: "Please override execute() method to intercept and handle errors like authorization or retry the request. Check the Wiki for more info. https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-do-i-implement-bearer-token-authentication-with-urlsession-on-the-swift-api-client")
+    @available(*, unavailable, message: "Please override execute() method to intercept and handle errors like authorization or retry the request. Check the Wiki for more info. https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-do-i-implement-bearer-token-authentication-with-urlsession-on-the-swift-api-client")
```
I don't know why these changes are only in my branch, they appeared when I generated samples

bernie-schelberg-mywave commented on samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift:
```diff
@@ -92,10 +92,6 @@ open class URLSessionRequestBuilder<T>: RequestBuilder<T> {
 
         originalRequest.httpMethod = method.rawValue
 
-        headers.forEach { key, value in
-            originalRequest.setValue(value, forHTTPHeaderField: key)
-        }
-
```
I don't know why these changes are only in my branch, they appeared when I generated samples

bernie-schelberg-mywave commented on samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift:
```diff
@@ -145,32 +141,13 @@ open class URLSessionRequestBuilder<T>: RequestBuilder<T> {
-                    }
+                apiResponseQueue.async {
+                    self.processRequestResponse(urlRequest: request, data: data, response: response, error: error, completion: completion)
+                    cleanupRequest()
```
I don't know why these changes are only in my branch, they appeared when I generated samples

bernie-schelberg-mywave commented on samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift:
```diff
@@ -145,32 +141,13 @@ open class URLSessionRequestBuilder<T>: RequestBuilder<T> {
-            if #available(iOS 11.0, macOS 10.13, macCatalyst 13.0, tvOS 11.0, watchOS 4.0, *) {
-                onProgressReady?(dataTask.progress)
-            }
+            onProgressReady?(dataTask.progress)
```
I don't know why these changes are only in my branch, they appeared when I generated samples

bernie-schelberg-mywave commented on samples/client/petstore/swift5/validation/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift:
```diff
@@ -218,10 +195,10 @@ open class URLSessionRequestBuilder<T>: RequestBuilder<T> {
             httpHeaders[key] = value
         }
-        for (key, value) in PetstoreClientAPI.customHeaders {
+        for (key, value) in headers {
```
I don't know why these changes are only in my branch, they appeared when I generated samples

bernie-schelberg-mywave commented on samples/openapi3/server/petstore/spring-boot-oneof/src/main/java/org/openapitools/model/Apple.java:
```diff
@@ -0,0 +1,128 @@
+   */
+  public Apple(Integer seeds) {
+    this.seeds = seeds;
+    this.fruitType = fruitType;
```
Initialising `fruitType` like this is weird, but is a separate issue I think. If a class implements an interface which has a oneOf with a discriminator, the discriminator property is only half considered "required". The template adds initialisation of the field, but omits the constructor parameter. Arguably, the field should be initialised with the correct discriminator value, it should not be a constructor parameter at all.

bernie-schelberg-mywave commented:
I've switched the branch from `master` to `7.0.x` on advice that this should be targeted for the 7.0.0 release. I added some better testing in the samples. Please review again and advise if there is anything else I can do to help move this forward.

cc @wing328 @jimschubert @cbornet @jmini @etherealjoy @spacether

effervescentia commented:
I was so excited to find this pull request 🙏 very eager to have this functionality added
unfortunately without it the `typescript-fetch` generator just creates uncompilable garbage (it's 99% there, just missing some type guards and trying to spread a non-object value into an object) so we aren't able to use it for the other endpoints until this feature has been released.

Thanks for your continued work on this @bernie-schelberg-mywave

effervescentia commented:
Hi @wing328 @jimschubert @cbornet @jmini @etherealjoy @spacether, I don't mean to be a pest, but is there a timeline we can expect to have this change merged?

Without this fix generators like the `typscript-fetch` generator will output uncompilable code with missing references and type-unsafe operations.
For example [this minimal schema](https://gist.github.com/effervescentia/543446af14c4a1dbf1c208f32cf7802c) will generate an SDK that fails to compile. I'm using [this configuration file](https://gist.github.com/effervescentia/ddfb5fc750e9f16fbaaa200050f7466b) which results in uncompilable code.

Each entry in the union creates [a file that looks like this](https://gist.github.com/effervescentia/259bf7dd49c7bccf1450e58efb571e44) (the output of the `'foo'` entry)

The union itself creates a [file with the invalid operations and missing references](https://gist.github.com/effervescentia/4ff7511b1f6a037f98d71b51dacb5012).

specifically:
1. the typeguards (`instanceOfTestControllerBadUnion200ResponseValueOneOf*`) do not exist so the import fails
2. the `*FromJSONTyped` functions have a string union return type (which is not accurate, they return objects) but are being used in a spread operation, which results in a compiler error

wing328 commented:
@bernie-schelberg-mywave thanks again for the PR. would you be able to resolve the merge conflicts for the last time? I'll merge it once the merge conflicts have been resolved.

wing328 commented:
as discussed the CI failure is not related to this change.

i'll look into the CI failures after merging this PR.

srix55 commented:
Hi @wing328 This fix isn't in 7.0.0-Beta? I am not seeing the changes there - it still says string required for me.

yarinsa commented:
Today when using discriminator a new type is generated. How can I use it and reference it to the original type? 
Meaning

```
enum System:
 AWS = 'aws'
 Azure = 'azure'

type Specs<S extends System> = { [key: string] : any; system: S }

type AWSSpecs = { account_id: number } & Specs<System.AWS>
type AZureSpecs = { client_id: number } & Specs<System.Azure>

type SystemSpecsMapper : Record<System, Specs<System>> = {
 [System.AWS]: AWSSpecs
 [System.Azure]: AzureSpecs
}

Request<S extends System> {
  name: S
  specs: SystemSpecsMapper<S>
}
```

Now when using discriminator with union (python) I am getting a whole new enum with a super long name,which has nothing to do with the original enum type
